src/control/server/ctl_smd_rpc.go

```diff
@@ -92,23 +92,37 @@ func (svc *ControlService) querySmdDevices(ctx context.Context, req *ctlpb.SmdQu
 			}
 		}
 
-		if !req.IncludeBioHealth {
-			continue
-		}
-
+		i := 0 // output index
 		for _, dev := range rResp.Devices {
-			/* Skip health query if the device is in "NEW" state */
-			if storage.NvmeDevState(dev.DevState).IsNew() {
-				continue
+			state := storage.NvmeDevState(dev.DevState)
+
+			if req.StateMask > 0 && req.StateMask&dev.DevState == 0 {
+				continue // skip device completely if mask doesn't match
 			}
-			health, err := ei.GetBioHealth(ctx, &ctlpb.BioHealthReq{
-				DevUuid: dev.Uuid,
-			})
-			if err != nil {
-				return errors.Wrapf(err, "device %s, states %q", dev,
-					storage.NvmeDevState(dev.DevState).String())
+
+			// skip health query if the device is in "NEW" state
+			if req.IncludeBioHealth && !state.IsNew() {
+				health, err := ei.GetBioHealth(ctx, &ctlpb.BioHealthReq{
+					DevUuid: dev.Uuid,
+				})
+				if err != nil {
+					return errors.Wrapf(err, "device %s, states %q", dev, state.String())
+				}
+				dev.Health = health
+			}
+
+			if req.StateMask > 0 {
+				// as mask is set and matches state, rewrite slice in place
+				rResp.Devices[i] = dev
+				i++
+			}
+		}
+		if req.StateMask > 0 {
```

Comment:
Doesn't matter too much since it's a uint at the moment, but I think it's more correct to use `req.StateMask != 0`, since the signedness (or lack thereof) isn't important for a bitmask.